pkg/client/snapshot.go

```diff
@@ -0,0 +1,183 @@
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	snapshotv1 "github.com/acorn-io/runtime/pkg/apis/snapshot.storage.k8s.io/v1"
+	"github.com/acorn-io/runtime/pkg/labels"
+	snapshot2 "github.com/acorn-io/runtime/pkg/snapshot"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+)
+
+var (
+	annotationAccessModes = "original-pvc-access-modes"
+	annotationResources   = "original-pvc-resource"
+	annotationVolumeName  = "original-pvc-volume-name"
+	annotationVolumeClass = "original-pvc-volume-class"
+)
+
+// SnapshotCreate creates a VolumeSnapshot resource in the cluster using the details of the given *PersistentVolumeClaim.
+// Returns a *VolumeSnapshot or error if one occurred.
+func (c *DefaultClient) SnapshotCreate(ctx context.Context, pvc *corev1.PersistentVolumeClaim) (*snapshotv1.VolumeSnapshot, error) {
+	err := snapshot2.CreateSnapshotClass(ctx, c.Client)
+	if err != nil {
+		return nil, err
+	}
+
+	name, ok := pvc.Labels["custom-name"]
+	if !ok {
+		name = pvc.Labels[labels.AcornPublicName] + "-" + strconv.FormatInt(time.Now().Unix(), 10)
+	}
+
+	accessModesJSON, err := json.Marshal(pvc.Spec.AccessModes)
+	if err != nil {
+		return nil, err
+	}
+
+	resourcesJSON, err := json.Marshal(pvc.Spec.Resources)
+	if err != nil {
+		return nil, err
+	}
+
+	snapshot := &snapshotv1.VolumeSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: pvc.Namespace,
+			Labels: map[string]string{
+				labels.AcornAppName:      pvc.Labels[labels.AcornAppName],
+				labels.AcornAppNamespace: pvc.Labels[labels.AcornAppNamespace],
+				labels.AcornPublicName:   pvc.Labels[labels.AcornPublicName],
+				labels.AcornManaged:      "true",
+			},
+			Annotations: map[string]string{
+				annotationAccessModes: string(accessModesJSON),
+				annotationResources:   string(resourcesJSON),
+				annotationVolumeName:  pvc.Labels[labels.AcornVolumeName],
+				annotationVolumeClass: pvc.Labels[labels.AcornVolumeClass],
+			},
+		},
+		Spec: snapshotv1.VolumeSnapshotSpec{
+			VolumeSnapshotClassName: pointer.String(snapshot2.ClassName),
+			Source: snapshotv1.VolumeSnapshotSource{
+				PersistentVolumeClaimName: pointer.String(pvc.Name),
+			},
+		},
+	}
+
+	return snapshot, c.Client.Create(ctx, snapshot)
+}
+
+// SnapshotList lists VolumeSnapshot. Returns []VolumeSnapshot or an error if one occurred.
+func (c *DefaultClient) SnapshotList(ctx context.Context) ([]snapshotv1.VolumeSnapshot, error) {
+	snapshots := &snapshotv1.VolumeSnapshotList{}
+	err := c.Client.List(ctx, snapshots)
```

Comment:
You should list the snapshot with namespace, otherwise it will return snapshot in all the projects